pkg/preference/implem.go

```diff
@@ -0,0 +1,564 @@
+	if err != nil {
+		return errors.Errorf("unable to set %q, something is wrong with odo, kindly raise an issue at https://github.com/redhat-developer/odo/issues/new?template=Bug.md", parameter)
+	}
+	return nil
+}
+
+// IsSet checks if the value is set in the preference
+func (c *preferenceInfo) IsSet(parameter string) bool {
+	return util.IsSet(c.OdoSettings, parameter)
+}
+
+// GetTimeout returns the value of Timeout from config
+// and if absent then returns default
+func (c *preferenceInfo) GetTimeout() int {
+	// default timeout value is 1
+	return util.GetIntOrDefault(c.OdoSettings.Timeout, DefaultTimeout)
+}
+
+// GetBuildTimeout gets the value set by BuildTimeout
+func (c *preferenceInfo) GetBuildTimeout() int {
+	// default timeout value is 300
+	return util.GetIntOrDefault(c.OdoSettings.BuildTimeout, DefaultBuildTimeout)
+}
+
+// GetPushTimeout gets the value set by PushTimeout
+func (c *preferenceInfo) GetPushTimeout() int {
+	// default timeout value is 1
+	return util.GetIntOrDefault(c.OdoSettings.PushTimeout, DefaultPushTimeout)
+}
+
+// GetRegistryCacheTime gets the value set by RegistryCacheTime
+func (c *preferenceInfo) GetRegistryCacheTime() int {
+	return util.GetIntOrDefault(c.OdoSettings.RegistryCacheTime, DefaultRegistryCacheTime)
+}
+
+// GetUpdateNotification returns the value of UpdateNotification from preferences
+// and if absent then returns default
+func (c *preferenceInfo) GetUpdateNotification() bool {
+	return util.GetBoolOrDefault(c.OdoSettings.UpdateNotification, true)
+}
+
+// GetEphemeralSourceVolume returns the value of ephemeral from preferences
+// and if absent then returns default
+func (c *preferenceInfo) GetEphemeralSourceVolume() bool {
+	return util.GetBoolOrDefault(c.OdoSettings.Ephemeral, DefaultEphemeralSettings)
+}
+
+// GetNamePrefix returns the value of Prefix from preferences
+// and if absent then returns default
+func (c *preferenceInfo) GetNamePrefix() string {
+	return util.GetStringOrEmpty(c.OdoSettings.NamePrefix)
+}
+
+// GetConsentTelemetry returns the value of ConsentTelemetry from preferences
+// and if absent then returns default
+// default value: false, consent telemetry is disabled by default
+func (c *preferenceInfo) GetConsentTelemetry() bool {
+	return util.GetBoolOrDefault(c.OdoSettings.ConsentTelemetry, DefaultConsentTelemetrySetting)
+}
+
+func (c *preferenceInfo) UpdateNotification() *bool {
+	return c.OdoSettings.UpdateNotification
+}
+
+func (c *preferenceInfo) NamePrefix() *string {
+	return c.OdoSettings.NamePrefix
+}
+
+func (c *preferenceInfo) Timeout() *int {
+	return c.OdoSettings.Timeout
+}
+
+func (c *preferenceInfo) BuildTimeout() *int {
+	return c.OdoSettings.BuildTimeout
+}
+
+func (c *preferenceInfo) PushTimeout() *int {
+	return c.OdoSettings.PushTimeout
+}
+
+func (c *preferenceInfo) EphemeralSourceVolume() *bool {
+	return c.OdoSettings.Ephemeral
+}
+
+func (c *preferenceInfo) ConsentTelemetry() *bool {
+	return c.OdoSettings.ConsentTelemetry
+}
+
```

Comment:
Why have these functions when you already have the other Getters on these properties? Why can't you directly access them? These additional methods are a little confusing to me.

Reply:
The other ones are returning the default value if the pointer value is nil or the dereferenced value otherwise, and these ones are returning the **raw** value (the pointer) - both are used in different places in the code.

We cannot access the values directly anymore because we want to access them through an interface (to be able to mock the full `Preference` object)